### PR TITLE
Support non-primitive types in callback parameters.

### PR DESCRIPTION
**Pull request by sfdctaka**
Non-primitive types can be passed to callback parameters. Following modifications were made:
* In Swift, `bind` methods constraint generic types to `Encodable` so that objects and primitives that can be JSON-ified be passed in.
* In Kotlin, annotation processor was updated to pass `JSONSerializable` object as is.
* Updated mocha tests.
* Updated demo-app.
* Updated binder tests in Swift.

Fixes https://github.com/salesforce/nimbus/issues/51.

### CODE REVIEW

File: platforms/apple/Sources/Nimbus/Callback.swift [hunk 1]
```diff
@@ -29,11 +29,21 @@ class Callback: Callable {
     }
 
     func call(args: [Any]) throws -> Any {
-        let jsonData = try JSONSerialization.data(withJSONObject: args, options: [])
-        let jsonString = String(data: jsonData, encoding: .utf8)
+        var jsonString: String = "[]"
+        if let encodables = args as? [EncodableValue] {
+            let jsonEncoder = JSONEncoder()
+            let jsonData = try? jsonEncoder.encode(encodables)
+            jsonString = String(data: jsonData!, encoding: .utf8)!
```
Comment (threeve) [inline platforms/apple/Sources/Nimbus/Callback.swift]: This `try?` followed by force-unwrapping the result will result in a misleading error being thrown if encoding fails. It would probably be better to use `try` and let the encoding error propagate up here.

File: platforms/apple/Sources/Nimbus/Callback.swift [hunk 2]
```diff
@@ -29,11 +29,21 @@ class Callback: Callable {
     }
 
     func call(args: [Any]) throws -> Any {
-        let jsonData = try JSONSerialization.data(withJSONObject: args, options: [])
-        let jsonString = String(data: jsonData, encoding: .utf8)
+        var jsonString: String = "[]"
+        if let encodables = args as? [EncodableValue] {
```
Comment (threeve) [inline platforms/apple/Sources/Nimbus/Callback.swift]: If the args are not encodable, what will happen here? It seems like the function will attempt to call through but will end up passing no arguments?

File: platforms/apple/Sources/Nimbus/Callback.swift [hunk 3]
```diff
@@ -29,11 +29,21 @@ class Callback: Callable {
...
+            let jsonData = try? jsonEncoder.encode(encodables)
+            jsonString = String(data: jsonData!, encoding: .utf8)!
+        }
+
         DispatchQueue.main.async {
             self.webView?.evaluateJavaScript("""
-            nimbus.callCallback('\(self.callbackId)', \(jsonString!));
+                var args = [];
+                var jsonArgs = \(jsonString);
+                jsonArgs.forEach(arg => {
```
Comment (threeve) [inline platforms/apple/Sources/Nimbus/Callback.swift]: you should be able to use Array.map here rather than imperatively pushing values onto a new array:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map

File: platforms/apple/Sources/NimbusTests/BinderTests.swift [hunk 4]
```diff
@@ -67,13 +67,14 @@ class BinderTests: XCTestCase {
 
     func testBindUnaryWithUnaryCallback() {
         binder.bind(BindTarget.unaryWithUnaryCallback, as: "")
+        //binder.bind(BindTarget.unaryWithUnaryCallbackThrows, as: "")
```
Comment (threeve) [inline platforms/apple/Sources/NimbusTests/BinderTests.swift]: remove?

File: platforms/android/nimbus/src/androidTest/java/com/salesforce/nimbus/MochaTests.kt [hunk 5]
```diff
@@ -23,7 +23,7 @@ import java.util.concurrent.TimeUnit
 @RunWith(AndroidJUnit4::class)
 class MochaTests {
 
-    data class MochaMessage(val stringField: String = "This is a string", val intField: Int = 42) : JSONSerializable {
+    data class MochaMessage(var stringField: String = "This is a string", var intField: Int = 42) : JSONSerializable {
```
Comment (threeve) [inline platforms/android/nimbus/src/androidTest/java/com/salesforce/nimbus/MochaTests.kt]: was this change necessary?

File: platforms/apple/Sources/Nimbus/Callback.swift (hunk 2 above)
Comment (threeve) [inline platforms/apple/Sources/Nimbus/Callback.swift]: There seems to be an implicit requirement here that `Callback` now must be called with `[EncodableValue]`, and this has resulted in quite a bit of churn in other code. I wonder if we might be able to express this with less churn by changing the expectation to `Callback` accepting an `[Encodable]` and adding a case to `EncodableValue` for handling `[Encodable]`? We might even be able to lift that requirement all the way to `Callable` which would be nice because then it's part of the actual interface and becomes an explicit constraint rather than implicit.

File: platforms/apple/Sources/NimbusTests/BinderTests.swift [hunk 6]
```diff
@@ -160,7 +163,7 @@ class BinderTests: XCTestCase {
         let callback: BindTarget.UnaryCallback = { value in
             expecter.fulfill()
         }
-        XCTAssertThrowsError(try binder.callable?.call(args: [42, make_callable(callback)]))
```
Comment (threeve) [inline platforms/apple/Sources/NimbusTests/BinderTests.swift]: all these changes to remove `make_callable` confuse me... can you shed some light on why you did this throughout the test cases?

File: platforms/apple/Sources/NimbusTests/BinderTests.swift [hunk 7]
```diff
@@ -429,8 +442,8 @@ enum BindError: Error {
 class BindTarget {
     private(set) var called = false
 
-    typealias UnaryCallback = (Int) -> Void
```
Comment (threeve) [inline platforms/apple/Sources/NimbusTests/BinderTests.swift]: I'm also a bit confused by this change, can you walk me through why these callback signatures changed?

File: test-www/test/globals.d.ts [hunk 8]
```diff
@@ -5,9 +5,20 @@
...
+}
+
 interface MochaTestBridge {
   ready(): void;
   sendMessage(name: string, includeParam: boolean): void;
 }
 
-declare var mochaTestBridge: MochaTestBridge;
\ No newline at end of file
+interface CallbackTestExtension {
```
Comment (threeve) [inline test-www/test/globals.d.ts]: the interfaces you added here could probably be placed in callback-encodable-tests.ts since they are specific to those test cases.

File: platforms/apple/Sources/Nimbus/Callback.swift (hunk 2 above)
Comment (sfdctaka) [reply platforms/apple/Sources/Nimbus/Callback.swift]: In `Binder.swift` callbacks are called with arguments(`args`) passed to them wrapped in `EncodableValue`. So in theory all elements in `args` should be `EncodableValue` but that is not easily perceived as the method signature of `call` which it inherits and defines from protocol `Callable`. If this assumption is valid, then, we can add an additional check such as if elements of `args` match with yielding array of JSON objects(after encoding elements of `args` individually instead of the whole array in one shot as I do it now).

File: platforms/android/nimbus/src/androidTest/java/com/salesforce/nimbus/MochaTests.kt (hunk 5 above)
Comment (sfdctaka) [reply platforms/android/nimbus/src/androidTest/java/com/salesforce/nimbus/MochaTests.kt]: I had to do this because for callback that takes to `MochaMessage`s as parameters I wanted to change the second one's property values so that they're different from the first `MochaMessage`.

File: platforms/apple/Sources/Nimbus/Callback.swift (hunk 2 above)
Comment (sfdctaka) [reply platforms/apple/Sources/Nimbus/Callback.swift]: Let me explore this.  Ping you tomorrow.

File: platforms/apple/Sources/NimbusTests/BinderTests.swift (hunk 6 above)
Comment (sfdctaka) [reply platforms/apple/Sources/NimbusTests/BinderTests.swift]: I removed `make_callable` because with it it produces `((Encodable)->()) throws -> Void` where as what I wanted was `(Encodable) -> ()`.

File: platforms/apple/Sources/NimbusTests/BinderTests.swift (hunk 7 above)
Comment (sfdctaka) [reply platforms/apple/Sources/NimbusTests/BinderTests.swift]: I changed this because in `Binder`'s `bind` method:
image
in line 61 I wrap `cb0` in `EncodableValue`.  

Then when the callback is called here and execution goes through line 51, then, type conversion to `A0` fails and was throwing `ParameterError`.
image

Same applies for `BinaryCallback` in `BindTarget`.

File: platforms/android/nimbus/src/androidTest/java/com/salesforce/nimbus/MochaTests.kt (hunk 5 above)
Comment (sfdctaka) [reply platforms/android/nimbus/src/androidTest/java/com/salesforce/nimbus/MochaTests.kt]: I got why you asked `val` vs `var`.  In the commit I had not included the file that requires this change.  It's coming in a follow-up commit.

File: platforms/apple/Sources/Nimbus/Callback.swift [hunk 9]
```diff
@@ -29,11 +29,26 @@ class Callback: Callable {
     }
 
     func call(args: [Any]) throws -> Any {
-        let jsonData = try JSONSerialization.data(withJSONObject: args, options: [])
-        let jsonString = String(data: jsonData, encoding: .utf8)
+        var jsonString: String = "[]"
+        if let encodables = args as? [Encodable] {
+            let jsonEncoder = JSONEncoder()
+            let encodableValues = encodables.map { encodable in
+                return EncodableValue.value(encodable)
```
Comment (sfdctaka) [inline platforms/apple/Sources/Nimbus/Callback.swift]: @threeve I tried to extend `EncodableValue` to be top-level class for `[Encodable]`.  However, I couldn't get `encode(...)` method to work and decided to wrap the elements of `args` here instead. Will ping you again tomorrow and see if there's a better approach.

Comment (threeve) [reply platforms/apple/Sources/Nimbus/Callback.swift]: It's possible to write such an `encode` method, let's work through it together.
